### ISSUE TITLE
Connect to what scoreboard?

### ISSUE THREAD

**Issue by MastaBaba** [stepmania/stepmania#1880]
Where do I discover what server I can connect to and how?

This is for the setting 'Network options' and, presumably, shared scoreboards.

**Comment by RhythmLunatic** [comment]
There are no servers left as far as I know, since http://stepmaniaonline.net/ doesn't even have account creation and leaderboards anymore

The feature is for SMOnline servers, the source code is here https://sourceforge.net/projects/smonline/ (+ there are some reimplementations on github)

You can host your own or modify it with whatever features you want.

**Comment by MastaBaba** [comment]
Thanks.

It seems that that bundle was built in C++. :/

Is there documentation on expected SMO endpoints and functionality?

**Comment by RhythmLunatic** [comment]
It is documented in /Docs/Devdocs/SMO_protocol-aj.txt

**Comment by MastaBaba** [comment]
Thanks again.

That lead me to this, which seems to be a more formatted text describing roughly the same:

https://stepmania-server.readthedocs.io/en/latest/protocol.html

If I'm reading this correctly, the Stepmania protocol allows for some elaborate multiplayer network games. Nice.

But, this is not really my cup of tea; I was hoping for an API with endpoints that takes POST and GET requests, sending and receiving JSON or similar. 
And, I was really only expecting leaderboards to synch.

With the server being written in C++, it's basically a black box to me. And I'd have no idea where to start hosting something written in C++.

What would be required to get a server up and running?

**Comment by concubidated** [comment]
That old server is very basic, there is a c# server that you can run with mono that is a bit more feature filled. I ran with it for many years.

https://github.com/concubidated/OpenSMO/tree/smo.net-live

Also there is a python3 server which works as well https://github.com/ningirsu/stepmania-server

**Comment by MastaBaba** [comment]
Thanks. 

For me, C# is not an improvement over C++. But, Python I probably would be able to manage, though this, too, is not what I normally work in.

What is not clear to me is how authentication of a particular user is supposed to work. That is, how is user registration and then authentication designed?

**Comment by RhythmLunatic** [comment]
Authentication and registration is left to the server to implement, SMO is just a protocol.

**Comment by MastaBaba** [comment]
I see. So, you're saying that authentication would have to somehow happen through the SMO protocol, yes?

**Comment by nico-abram** [comment]
iirc the smo protocol lets you send a username+password pair ingame. *How* that gets authenthicated is entirely up to the server (Iirc I've seen at least one implementation that just let you login without verifying anything).

Also, @ 
> But, this is not really my cup of tea; I was hoping for an API with endpoints that takes POST and GET requests, sending and receiving JSON or similar.
> And, I was really only expecting leaderboards to synch.

I'm guessing what you had in mind was more similar to having a leaderboards server/website that just exposes some kind of web api (i.e rest with json or w.e) and then recieves stuff from the client asynchronously (Like, whenever the client gets a new score, it sends it to the leaderboards web server), and not actual real-time multiplayer? The smo protocol is afaik mainly just for playing the same song with friends at roughly the same time. The fact that the server acts as the intermediary between players allows it to store scores, but only scores which are played while connected to the multiplayer server and playing in a room. That is, afaik, how smo (stepmaniaonline) works/worked. 

If what I described in the previously was what you had in mind it sounds a lot more similar to what a fork of stepmania (Called etterna and focused on 4key keyboard play) does. However, the situation there is a bit weird since there are essentially two pretty much unrelated servers (The "multiplayer" server, and the "stats tracking" web server with it's rest API) with different ip addresses (Which can be a bit confusing for users)

**Comment by MastaBaba** [comment]
@nico-abram Yes, indeed. I was not thinking of multiplayer. Only of synchronising scores.

I just tried installing Etterna, thanks for the reference, but it failed on MacOS, claiming some library was missing. :(

**Comment by RhythmLunatic** [comment]
Chrispable's fork (and my fork, being based off his) has scoreboard submission.

There is no authentication or verification, it's probably meant for arcades only.

**Comment by MastaBaba** [comment]
Thanks @RhythmLunatic . I'll see if I can get your fork up and running.

Where are the scoreboards maintained?

**Comment by RhythmLunatic** [comment]
There are none. It's to make your own. I think pk has his own but it's not public, maybe you can ask him

Check https://github.com/RhythmLunatic/stepmania/blob/starworlds/src/ProfileManager.cpp#L959-L1039 to see what gets submitted.

My buddy dj505 has a leaderboard software for pump, it's possible that it could be modified to support sm score submission but I don't care to do that myself https://github.com/dj505/ArrowTracker

**Comment by MastaBaba** [comment]
Thanks for the insight, @RhythmLunatic . I'll first see if I can get a Linux build of your fork up and running.
